ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+   * ClusterInitializationFailedError, which holds a message and the current Cluster object (if one
+   * existed at the time of failure).
+   */
+  public async run(): Promise<Cluster> {
+    if (this.isInitializing) {
+      throw new Error('Initialization is already in progress');
+    }
+    this.initializeStartTime = Date.now();
+
+    return new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject as (error: Error) => {};
+      this.poll();
+    }) as Promise<Cluster>;
+  }
+
+  private async poll() {
+    // Overall strategy: continue polling the get-cluster endpoint, with capped exponential backoff,
+    // until we either reach our goal state (a RUNNING cluster) or run up against the overall
+    // timeout threshold.
+    //
+    // Certain cluster states require active intervention, such as deleting or resuming the cluster;
+    // these are handled within the the polling loop.
+    if (this.abortSignal && this.abortSignal.aborted) {
+      // We'll bail out early if an abort signal was triggered while waiting for the poll cycle.
+      return this.reject(
+        new ClusterInitializationFailedError('Request was aborted.', this.currentCluster));
+    }
+    if (Date.now() - this.initializeStartTime > this.overallTimeout) {
+      return this.reject(
+        new ClusterInitializationFailedError(
+          'Initialization attempt took longer than the max time allowed.',
+          this.currentCluster));
+    }
+
+    // Fetch the current cluster status, with some graceful error handling for NOT_FOUND response
+    // and abort signals.
+    try {
+      this.currentCluster = await this.getCluster();
+      this.onStatusUpdate(this.currentCluster.status);
+    } catch (e) {
+      if (isAbortError(e)) {
+        return this.reject(
+          new ClusterInitializationFailedError('Abort signal received during cluster API call',
+            this.currentCluster));
+      } else if (this.isNotFoundError(e)) {
+        // A not-found error is somewhat expected, if a cluster has recently been deleted or
+        // hasn't been created yet.
+        this.currentCluster = null;
+      } else {
+        this.handleUnknownError(e);
+        if (this.tooManyServerErrors()) {
+          return this.reject(
+            new ExceededErrorCountError('Reached max server error count', this.currentCluster));
+        }
+      }
+    }
+
+    // Attempt to take the appropriate next action given the current cluster status.
+    try {
+      if (this.currentCluster === null) {
+        await this.createCluster();
+      } else if (this.isClusterStopped()) {
+        await this.resumeCluster();
+      } else if (this.isClusterErrored()) {
+        // If cluster is in error state, delete it so it can be re-created at the next poll loop.
+        reportError(
+          `Cluster ${this.currentCluster.clusterNamespace}/${this.currentCluster.clusterName}` +
```

Comment:
nit: exposing a property for the full cluster path/fqn would be cleaner than doing it here I think.